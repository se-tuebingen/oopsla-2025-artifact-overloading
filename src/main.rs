use overloading::compile_expr;
use overloading::syntax::parse_expr;

use clap::Parser;

#[derive(Parser)]
#[command(name = "overloading")]
struct Args {
    /// Input file to process
    filename: String,

    /// Use short output format
    #[arg(short, long)]
    short: bool,
}

/// Main entry point for the native CLI.
fn main() {
    let args = Args::parse();

    let content = std::fs::read_to_string(&args.filename).expect("Failed to read file");
    match parse_expr(&content) {
        Ok(expr) => {
            println!("{}", compile_expr(&expr, args.short));
        }
        Err(err) => {
            panic!("Failed to parse expression: {}", err);
        }
    }
}

/// Snapshot tests for the compiler examples.
#[cfg(test)]
#[test]
fn test_compiler_examples() {
    use insta::{assert_json_snapshot, glob, with_settings};

    glob!("../", "examples/*.over", |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let output = match parse_expr(&input) {
            Ok(expr) => overloading::compile_expr_json(&expr, true),
            Err(err) => overloading::json::JsonProcessingResult::InferenceError {
                expression: input,
                error: err,
            },
        };

        // Use the filename (without extension) as snapshot suffix
        let filename = path.file_stem().unwrap().to_str().unwrap();
        with_settings!({
            snapshot_path => "../examples/snapshots",
            snapshot_suffix => filename,
            sort_maps => true,
        }, {
            assert_json_snapshot!(output);
        });
    });
}
