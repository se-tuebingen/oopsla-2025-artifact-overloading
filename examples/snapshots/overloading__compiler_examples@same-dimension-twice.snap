---
source: src/main.rs
expression: output
input_file: examples/same-dimension-twice.over
---
{
  "status": "success",
  "expression": "assume addInt: (Int, Int) => Int\nassume addString: (String, String) => String\nλ x => let add_b = a‹addInt, addString›\nlet inner = add_b(x, x)\nlet add_a = a‹addInt, addString›\nadd_a(0, inner) => \n  let add_b = a‹addInt, addString›\n  let inner = add_b(x, x)\n  let add_a = a‹addInt, addString›\n  add_a(0, inner)",
  "inferred_type": "(x_0) => app_4",
  "worlds": {
    "type": "one"
  },
  "solutions": [
    [
      "a₁"
    ]
  ],
  "constraints": [
    {
      "from": "(Int, Int) => Int",
      "choices": [
        "a₁"
      ],
      "to": "over_a_1"
    },
    {
      "from": "(String, String) => String",
      "choices": [
        "a₂"
      ],
      "to": "over_a_1"
    },
    {
      "from": "over_a_1",
      "choices": [],
      "to": "(x_0, x_0) => app_2"
    },
    {
      "from": "(Int, Int) => Int",
      "choices": [
        "a₁"
      ],
      "to": "over_a_3"
    },
    {
      "from": "(String, String) => String",
      "choices": [
        "a₂"
      ],
      "to": "over_a_3"
    },
    {
      "from": "over_a_3",
      "choices": [],
      "to": "(Int, app_2) => app_4"
    }
  ],
  "bounds": [
    {
      "variable": "app_2",
      "lower": [
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "a₂"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "a₂"
          ]
        }
      ]
    },
    {
      "variable": "app_4",
      "lower": [
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "a₂"
          ]
        }
      ],
      "upper": []
    },
    {
      "variable": "over_a_1",
      "lower": [
        {
          "type_name": "(Int, Int) => Int",
          "choices": [
            "a₁"
          ]
        },
        {
          "type_name": "(String, String) => String",
          "choices": [
            "a₂"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "(x_0, x_0) => app_2",
          "choices": []
        }
      ]
    },
    {
      "variable": "over_a_3",
      "lower": [
        {
          "type_name": "(Int, Int) => Int",
          "choices": [
            "a₁"
          ]
        },
        {
          "type_name": "(String, String) => String",
          "choices": [
            "a₂"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "(Int, app_2) => app_4",
          "choices": []
        }
      ]
    },
    {
      "variable": "x_0",
      "lower": [],
      "upper": [
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "a₂"
          ]
        }
      ]
    }
  ],
  "errors": [
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "a₁",
          "a₂"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "a₂"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "a₂",
          "a₁"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    }
  ]
}
